File: automate_sd.py
#!/usr/bin/env python3
"""
Stable Diffusion WebUI Automation Script

This script provides programmatic access to the Stable Diffusion WebUI API
for automated image generation without needing the GUI.

Usage:
    python automate_sd.py --prompt "a beautiful landscape" --output output.png
    python automate_sd.py --prompt "portrait of a person" --steps 50 --cfg 7.5
"""

import argparse
import base64
import json
import os
import sys
import time
from pathlib import Path
from typing import Optional

import requests


class StableDiffusionAPI:
    """Python wrapper for Stable Diffusion WebUI API"""
    
    def __init__(self, base_url: str = "http://127.0.0.1:7860"):
        self.base_url = base_url
        self.api_endpoint = f"{base_url}/sdapi/v1"
        
    def wait_for_api(self, timeout: int = 60) -> bool:
        """Wait for the WebUI API to be ready"""
        print(f"Waiting for WebUI at {self.base_url}...")
        start_time = time.time()
        while time.time() - start_time < timeout:
            try:
                response = requests.get(f"{self.api_endpoint}/options", timeout=5)
                if response.status_code == 200:
                    print("WebUI API is ready!")
                    return True
            except requests.exceptions.RequestException:
                pass
            time.sleep(2)
        print("Timeout waiting for WebUI")
        return False
    
    def txt2img(
        self,
        prompt: str,
        negative_prompt: str = "",
        steps: int = 20,
        cfg_scale: float = 7.0,
        width: int = 512,
        height: int = 512,
        seed: int = -1,
        batch_size: int = 1,
        sampler_name: str = "Euler a"
    ) -> list:
        """
        Generate images from text prompt
        
        Args:
            prompt: Positive prompt describing what to generate
            negative_prompt: What to avoid in the image
            steps: Number of denoising steps (higher = better quality, slower)
            cfg_scale: Classifier free guidance scale (7 is good default)
            width: Image width
            height: Image height
            seed: Random seed (-1 for random)
            batch_size: Number of images to generate
            sampler_name: Sampling method
            
        Returns:
            List of base64 encoded images
        """
        payload = {
            "prompt": prompt,
            "negative_prompt": negative_prompt,
            "steps": steps,
            "cfg_scale": cfg_scale,
            "width": width,
            "height": height,
            "seed": seed,
            "batch_size": batch_size,
            "sampler_name": sampler_name,
        }
        
        print(f"Generating: '{prompt}' (steps={steps}, cfg={cfg_scale})")
        response = requests.post(
            f"{self.api_endpoint}/txt2img",
            json=payload,
            timeout=300  # 5 minute timeout for generation
        )
        
        if response.status_code != 200:
            raise Exception(f"API error: {response.status_code} - {response.text}")
            
        result = response.json()
        return result.get("images", [])
    
    def img2img(
        self,
        prompt: str,
        image_path: str,
        negative_prompt: str = "",
        steps: int = 20,
        cfg_scale: float = 7.0,
        denoising_strength: float = 0.75,
        seed: int = -1,
    ) -> list:
        """Generate images from image + text prompt"""
        # Read and encode image
        with open(image_path, "rb") as f:
            image_bytes = f.read()
        image_b64 = base64.b64encode(image_bytes).decode("utf-8")
        
        payload = {
            "prompt": prompt,
            "negative_prompt": negative_prompt,
            "init_images": [image_b64],
            "steps": steps,
            "cfg_scale": cfg_scale,
            "denoising_strength": denoising_strength,
            "seed": seed,
        }
        
        print(f"Img2Img: '{prompt}' from {image_path}")
        response = requests.post(
            f"{self.api_endpoint}/img2img",
            json=payload,
            timeout=300
        )
        
        if response.status_code != 200:
            raise Exception(f"API error: {response.status_code}")
            
        result = response.json()
        return result.get("images", [])
    
    def save_image(self, base64_data: str, output_path: str) -> str:
        """Save base64 image to file"""
        image_bytes = base64.b64decode(base64_data)
        with open(output_path, "wb") as f:
            f.write(image_bytes)
        return output_path
    
    def get_models(self) -> list:
        """Get available models"""
        response = requests.get(f"{self.api_endpoint}/sd-models")
        return response.json()
    
    def get_samplers(self) -> list:
        """Get available samplers"""
        response = requests.get(f"{self.api_endpoint}/samplers")
        return response.json()


def generate_batch(prompts: list, output_dir: str = "outputs", **kwargs):
    """Generate multiple images from a list of prompts"""
    os.makedirs(output_dir, exist_ok=True)
    api = StableDiffusionAPI()
    
    if not api.wait_for_api():
        print("Failed to connect to WebUI")
        return
    
    for i, prompt in enumerate(prompts):
        print(f"\n[{i+1}/{len(prompts)}] {prompt}")
        try:
            images = api.txt2img(prompt, **kwargs)
            for j, img_data in enumerate(images):
                output_path = os.path.join(output_dir, f"gen_{i}_{j}.png")
                api.save_image(img_data, output_path)
                print(f"  Saved: {output_path}")
        except Exception as e:
            print(f"  Error: {e}")


def main():
    parser = argparse.ArgumentParser(description="Stable Diffusion WebUI Automation")
    parser.add_argument("--prompt", type=str, required=True, help="Text prompt")
    parser.add_argument("--negative", type=str, default="", help="Negative prompt")
    parser.add_argument("--output", type=str, default="output.png", help="Output file")
    parser.add_argument("--steps", type=int, default=20, help="Number of steps")
    parser.add_argument("--cfg", type=float, default=7.0, help="CFG scale")
    parser.add_argument("--width", type=int, default=512, help="Image width")
    parser.add_argument("--height", type=int, default=512, help="Image height")
    parser.add_argument("--seed", type=int, default=-1, help="Seed (-1 for random)")
    parser.add_argument("--sampler", type=str, default="Euler a", help="Sampler name")
    parser.add_argument("--url", type=str, default="http://127.0.0.1:7860", help="WebUI URL")
    parser.add_argument("--img2img", type=str, help="Input image for img2img")
    parser.add_argument("--denoise", type=float, default=0.75, help="Denoising strength for img2img")
    
    args = parser.parse_args()
    
    api = StableDiffusionAPI(args.url)
    
    if not api.wait_for_api():
        print("Error: Could not connect to WebUI. Is it running?")
        sys.exit(1)
    
    try:
        if args.img2img:
            images = api.img2img(
                prompt=args.prompt,
                image_path=args.img2img,
                negative_prompt=args.negative,
                steps=args.steps,
                cfg_scale=args.cfg,
                denoising_strength=args.denoise,
                seed=args.seed
            )
        else:
            images = api.txt2img(
                prompt=args.prompt,
                negative_prompt=args.negative,
                steps=args.steps,
                cfg_scale=args.cfg,
                width=args.width,
                height=args.height,
                seed=args.seed,
                sampler_name=args.sampler
            )
        
        for i, img_data in enumerate(images):
            output_path = args.output if i == 0 else args.output.replace(".png", f"_{i}.png")
            api.save_image(img_data, output_path)
            print(f"Saved: {output_path}")
            
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: automate_sd.js
/**
 * Stable Diffusion WebUI Automation - JavaScript/Node.js Version
 * 
 * This script provides programmatic access to the Stable Diffusion WebUI API
 * for automated image generation in Node.js/JavaScript environments.
 * 
 * Usage:
 *   node automate_sd.js --prompt "your prompt" --output image.png
 *   node automate_sd.js --prompt "landscape" --steps 30 --width 1024 --height 576
 * 
 * Or use as a module:
 *   const { StableDiffusionAPI } = require('./automate_sd.js');
 */

const http = require('http');
const fs = require('fs');
const path = require('path');

// Configuration
const DEFAULT_URL = process.env.SD_WEBUI_URL || 'http://127.0.0.1:7860';
const API_BASE = '/sdapi/v1';

/**
 * Make HTTP request to WebUI API
 */
function apiRequest(endpoint, method = 'GET', data = null) {
    return new Promise((resolve, reject) => {
        const url = new URL(endpoint, DEFAULT_URL);
        const options = {
            hostname: url.hostname,
            port: url.port || 7860,
            path: url.pathname,
            method: method,
            headers: {
                'Content-Type': 'application/json'
            }
        };

        const req = http.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(body));
                } catch (e) {
                    resolve(body);
                }
            });
        });

        req.on('error', reject);
        
        if (data) {
            req.write(JSON.stringify(data));
        }
        req.end();
    });
}

/**
 * Stable Diffusion WebUI API Wrapper
 */
class StableDiffusionAPI {
    constructor(baseUrl = DEFAULT_URL) {
        this.baseUrl = baseUrl;
    }

    /**
     * Wait for WebUI API to be ready
     */
    async waitForApi(timeout = 60) {
        console.log(`Waiting for WebUI at ${this.baseUrl}...`);
        const start = Date.now();
        
        while (Date.now() - start < timeout * 1000) {
            try {
                await this.getOptions();
                console.log('WebUI API is ready!');
                return true;
            } catch (e) {
                await new Promise(r => setTimeout(r, 2000));
            }
        }
        console.log('Timeout waiting for WebUI');
        return false;
    }

    /**
     * Get available options
     */
    async getOptions() {
        return apiRequest(`${API_BASE}/options`, 'GET');
    }

    /**
     * Get available models
     */
    async getModels() {
        return apiRequest(`${API_BASE}/sd-models`, 'GET');
    }

    /**
     * Get available samplers
     */
    async getSamplers() {
        return apiRequest(`${API_BASE}/samplers`, 'GET');
    }

    /**
     * Generate images from text prompt (txt2img)
     */
    async txt2img(options = {}) {
        const {
            prompt = '',
            negative_prompt = '',
            steps = 20,
            cfg_scale = 7.0,
            width = 512,
            height = 512,
            seed = -1,
            batch_size = 1,
            sampler_name = 'Euler a'
        } = options;

        console.log(`Generating: '${prompt}' (steps=${steps}, cfg=${cfg_scale})`);
        
        const payload = {
            prompt,
            negative_prompt,
            steps,
            cfg_scale,
            width,
            height,
            seed,
            batch_size,
            sampler_name
        };

        const result = await apiRequest(`${API_BASE}/txt2img`, 'POST', payload);
        return result.images || [];
    }

    /**
     * Generate images from image + text prompt (img2img)
     */
    async img2img(options = {}) {
        const {
            prompt = '',
            negative_prompt = '',
            init_images = [],
            steps = 20,
            cfg_scale = 7.0,
            denoising_strength = 0.75,
            seed = -1,
            sampler_name = 'Euler a'
        } = options;

        console.log(`Img2Img: '${prompt}'`);
        
        const payload = {
            prompt,
            negative_prompt,
            init_images,
            steps,
            cfg_scale,
            denoising_strength,
            seed,
            sampler_name
        };

        const result = await apiRequest(`${API_BASE}/img2img`, 'POST', payload);
        return result.images || [];
    }

    /**
     * Read and encode image to base64
     */
    readImageAsBase64(imagePath) {
        const buffer = fs.readFileSync(imagePath);
        return buffer.toString('base64');
    }

    /**
     * Save base64 image to file
     */
    saveImage(base64Data, outputPath) {
        const buffer = Buffer.from(base64Data, 'base64');
        const dir = path.dirname(outputPath);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }
        fs.writeFileSync(outputPath, buffer);
        return outputPath;
    }

    /**
     * Generate and save image in one call
     */
    async generate(options = {}) {
        const { output = 'output.png', ...txt2imgOptions } = options;
        const images = await this.txt2img(txt2imgOptions);
        if (images.length > 0) {
            this.saveImage(images[0], output);
            console.log(`Saved: ${output}`);
            return output;
        }
        throw new Error('No images generated');
    }
}

// CLI Interface
async function main() {
    const args = process.argv.slice(2);
    const options = {
        prompt: '',
        negative: '',
        output: 'output.png',
        steps: 20,
        cfg: 7.0,
        width: 512,
        height: 512,
        seed: -1,
        sampler: 'Euler a',
        url: DEFAULT_URL,
        img2img: null,
        denoise: 0.75
    };

    // Parse arguments
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        switch (arg) {
            case '--prompt':
            case '-p':
                options.prompt = args[++i];
                break;
            case '--negative':
            case '-n':
                options.negative = args[++i];
                break;
            case '--output':
            case '-o':
                options.output = args[++i];
                break;
            case '--steps':
            case '-s':
                options.steps = parseInt(args[++i]);
                break;
            case '--cfg':
                options.cfg = parseFloat(args[++i]);
                break;
            case '--width':
                options.width = parseInt(args[++i]);
                break;
            case '--height':
                options.height = parseInt(args[++i]);
                break;
            case '--seed':
                options.seed = parseInt(args[++i]);
                break;
            case '--sampler':
                options.sampler = args[++i];
                break;
            case '--url':
                options.url = args[++i];
                break;
            case '--img2img':
            case '-i':
                options.img2img = args[++i];
                break;
            case '--denoise':
                options.denoise = parseFloat(args[++i]);
                break;
            case '--help':
            case '-h':
                console.log(`
Stable Diffusion WebUI Automation - JavaScript Version

Usage: node automate_sd.js [options]

Options:
  -p, --prompt <text>     Text prompt (required)
  -n, --negative <text>   Negative prompt
  -o, --output <file>    (default: output Output filename.png)
  -s, --steps <num>      Number of steps (default: 20)
  --cfg <num>            CFG scale (default: 7.0)
  --width <num>          Image width (default: 512)
  --height <num>         Image height (default: 512)
  --seed <num>           Seed (-1 for random)
  --sampler <name>       Sampler (default: Euler a)
  --url <url>            WebUI URL (default: http://127.0.0.1:7860)
  -i, --img2img <file>   Input image for img2img
  --denoise <num>        Denoising strength (default: 0.75)
  -h, --help            Show this help

Examples:
  node automate_sd.js -p "a cat" -o cat.png
  node automate_sd.js --prompt "landscape" --steps 50 --width 1024 --height 576
  node automate_sd.js -i input.png --prompt "oil painting" --denoise 0.6
                `);
                process.exit(0);
        }
    }

    if (!options.prompt) {
        console.error('Error: --prompt is required');
        console.log('Use --help for usage information');
        process.exit(1);
    }

    const api = new StableDiffusionAPI(options.url);
    
    if (!await api.waitForApi()) {
        console.error('Error: Could not connect to WebUI');
        process.exit(1);
    }

    try {
        if (options.img2img) {
            // Image-to-image mode
            const imageB64 = api.readImageAsBase64(options.img2img);
            const images = await api.img2img({
                prompt: options.prompt,
                negative_prompt: options.negative,
                init_images: [imageB64],
                steps: options.steps,
                cfg_scale: options.cfg,
                denoising_strength: options.denoise,
                seed: options.seed,
                sampler_name: options.sampler
            });
            if (images.length > 0) {
                api.saveImage(images[0], options.output);
                console.log(`Saved: ${options.output}`);
            }
        } else {
            // Text-to-image mode
            await api.generate({
                prompt: options.prompt,
                negative_prompt: options.negative,
                output: options.output,
                steps: options.steps,
                cfg_scale: options.cfg,
                width: options.width,
                height: options.height,
                seed: options.seed,
                sampler_name: options.sampler
            });
        }
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

// Export for use as module
module.exports = { StableDiffusionAPI, apiRequest };

// Run CLI if executed directly
if (require.main === module) {
    main();
}
